Add support for more browser.extension incognito Web Extension APIs.

xeenon opened the pull request:
#### 2713dbbe432dc19ca160330e3a76ac4296021d7c
<pre>
Add support for more browser.extension incognito Web Extension APIs.
<a href="https://bugs.webkit.org/show_bug.cgi?id=262848">https://bugs.webkit.org/show_bug.cgi?id=262848</a>
rdar://problem/116629181

Reviewed by Brady Eidson.

- Fixes some more cases that needed to check if content can be injected in private tabs.
- Added support for the incognito APIs on browser.extension.
- Added support for file scheme API that always return false.
- Added better window and tab test support, specifically for closing windows and tabs.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIExtensionCocoa.mm: Added.
(WebKit::WebExtensionContext::extensionIsAllowedIncognitoAccess):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::setHasAccessInPrivateBrowsing): Added.
(WebKit::WebExtensionContext::removeGrantedPermissions): Return a bool.
(WebKit::WebExtensionContext::removeGrantedPermissionMatchPatterns): Ditto.
(WebKit::WebExtensionContext::removeDeniedPermissions): Ditto.
(WebKit::WebExtensionContext::removeDeniedPermissionMatchPatterns): Ditto.
(WebKit::WebExtensionContext::removePermissions): Ditto.
(WebKit::WebExtensionContext::removePermissionMatchPatterns): Ditto.
(WebKit::WebExtensionContext::removeExpired): Call updateInjectedContent to remove scripts from expired patterns.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::addUserContentController): Track private controllers too. Only add injected content
if the extension has access to that controller.
(WebKit::WebExtensionController::removeUserContentController): Remove from private controller set.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const): Infer return type.
(WebKit::WebExtensionContext::pageListensForEvent const): Check hasAccessInPrivateBrowsing().
(WebKit::WebExtensionContext::processes const): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::setHasAccessInPrivateBrowsing): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::allUserContentControllers const):
(WebKit::WebExtensionController::allPrivateUserContentControllers const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:
(WebKit::WebExtensionAPIExtension::getURL):
(WebKit::WebExtensionAPIExtension::isInIncognitoContext): Added.
(WebKit::WebExtensionAPIExtension::isAllowedFileSchemeAccess): Added.
(WebKit::WebExtensionAPIExtension::isAllowedIncognitoAccess): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIExtension.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIExtension.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIExtension.mm:
(TestWebKitAPI::TEST): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Use class tab type.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initForExtension:]): Default to an empty yield message.
(-[TestWebExtensionManager closeWindow:]): Close all tabs, set defaultWindow to the next window and focus it.
(-[TestWebExtensionManager run]): Clear yield message.
(-[TestWebExtensionManager runForTimeInterval:]): Ditto.
(-[TestWebExtensionManager _webExtensionController:recordTestYieldedWithMessage:andSourceURL:lineNumber:forExtensionContext:]):
Always set a non-nil yield message.
(-[TestWebExtensionTab initWithWindow:extensionController:]): Use class type.
(-[TestWebExtensionTab activateForWebExtensionContext:completionHandler:]): Ditto.
(-[TestWebExtensionTab isSelectedForWebExtensionContext:]): Ditto.
(-[TestWebExtensionTab closeForWebExtensionContext:completionHandler:]): Ditto.
(-[TestWebExtensionWindow tabs]): Ditto.
(-[TestWebExtensionWindow setTabs:]): Close and open tabs as needed.
(-[TestWebExtensionWindow openNewTabAtIndex:]): Use class type.
(-[TestWebExtensionWindow closeTab:]): Call closeTab:windowIsClosing:.
(-[TestWebExtensionWindow closeTab:windowIsClosing:]): Added. Close the web view. Update activeTab if needed.
(-[TestWebExtensionWindow replaceTab:withTab:]): Close the web view on the replaced tab.
(-[TestWebExtensionWindow moveTab:toIndex:]): Use class type.

Canonical link: <a href="https://commits.webkit.org/269087@main">https://commits.webkit.org/269087@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8c19bfd204a5eff548e0956cabea04fb95f99fae

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21586 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/26/builds/21898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/22635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/23449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/19994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/21829 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/23/builds/25214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/22131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/23449 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/21813 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/23/builds/25214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/22635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/24299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/23/builds/25214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/22635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/24299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/23/builds/25214 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/22635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/24299 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/20249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/22131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/19570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/22635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/23804 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2673 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/20156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
```diff
@@ -99,11 +99,10 @@ class WebExtensionController : public API::ObjectImpl<API::Object::Type::WebExte
-
-    // Excludes private browsing content controllers.
     UserContentControllerProxySet allNonPrivateUserContentControllers() const { return m_allNonPrivateUserContentControllers; }
+    UserContentControllerProxySet allPrivateUserContentControllers() const { return m_allPrivateUserContentControllers; }
```
I like having all three defined! Definitely avoids confusion with what "normal" user content controllers means.

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:
```diff
@@ -46,12 +51,36 @@
+void WebExtensionAPIExtension::isAllowedFileSchemeAccess(Ref<WebExtensionCallbackHandler>&& callback)
+{
+    // Documentation: https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/extension/isAllowedFileSchemeAccess
+    // FIXME: rdar://problem/58428135 Consider allowing file URL access if the user opted in explicitly in some way.
```
Should this be a Bugzilla bug instead?

xeenon replied:
I kept it as a radar since we already have it an it is more of a policy decision that will need addressed before it can be fixed, which is more internal centric.

xeenon committed:
Add support for more browser.extension incognito Web Extension APIs.
https://bugs.webkit.org/show_bug.cgi?id=262848
rdar://problem/116629181

Reviewed by Brady Eidson.

- Fixes some more cases that needed to check if content can be injected in private tabs.
- Added support for the incognito APIs on browser.extension.
- Added support for file scheme API that always return false.
- Added better window and tab test support, specifically for closing windows and tabs.

* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIExtensionCocoa.mm: Added.
(WebKit::WebExtensionContext::extensionIsAllowedIncognitoAccess):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::setHasAccessInPrivateBrowsing): Added.
(WebKit::WebExtensionContext::removeGrantedPermissions): Return a bool.
(WebKit::WebExtensionContext::removeGrantedPermissionMatchPatterns): Ditto.
(WebKit::WebExtensionContext::removeDeniedPermissions): Ditto.
(WebKit::WebExtensionContext::removeDeniedPermissionMatchPatterns): Ditto.
(WebKit::WebExtensionContext::removePermissions): Ditto.
(WebKit::WebExtensionContext::removePermissionMatchPatterns): Ditto.
(WebKit::WebExtensionContext::removeExpired): Call updateInjectedContent to remove scripts from expired patterns.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm:
(WebKit::WebExtensionController::addUserContentController): Track private controllers too. Only add injected content
if the extension has access to that controller.
(WebKit::WebExtensionController::removeUserContentController): Remove from private controller set.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.cpp:
(WebKit::WebExtensionContext::parameters const): Infer return type.
(WebKit::WebExtensionContext::pageListensForEvent const): Check hasAccessInPrivateBrowsing().
(WebKit::WebExtensionContext::processes const): Ditto.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::setHasAccessInPrivateBrowsing): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionController.h:
(WebKit::WebExtensionController::allUserContentControllers const):
(WebKit::WebExtensionController::allPrivateUserContentControllers const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIExtensionCocoa.mm:
(WebKit::WebExtensionAPIExtension::getURL):
(WebKit::WebExtensionAPIExtension::isInIncognitoContext): Added.
(WebKit::WebExtensionAPIExtension::isAllowedFileSchemeAccess): Added.
(WebKit::WebExtensionAPIExtension::isAllowedIncognitoAccess): Added.
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIExtension.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIExtension.idl:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIExtension.mm:
(TestWebKitAPI::TEST): Added.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPITabs.mm:
(TestWebKitAPI::TEST): Use class tab type.
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.h:
* Tools/TestWebKitAPI/cocoa/WebExtensionUtilities.mm:
(-[TestWebExtensionManager initForExtension:]): Default to an empty yield message.
(-[TestWebExtensionManager closeWindow:]): Close all tabs, set defaultWindow to the next window and focus it.
(-[TestWebExtensionManager run]): Clear yield message.
(-[TestWebExtensionManager runForTimeInterval:]): Ditto.
(-[TestWebExtensionManager _webExtensionController:recordTestYieldedWithMessage:andSourceURL:lineNumber:forExtensionContext:]):
Always set a non-nil yield message.
(-[TestWebExtensionTab initWithWindow:extensionController:]): Use class type.
(-[TestWebExtensionTab activateForWebExtensionContext:completionHandler:]): Ditto.
(-[TestWebExtensionTab isSelectedForWebExtensionContext:]): Ditto.
(-[TestWebExtensionTab closeForWebExtensionContext:completionHandler:]): Ditto.
(-[TestWebExtensionWindow tabs]): Ditto.
(-[TestWebExtensionWindow setTabs:]): Close and open tabs as needed.
(-[TestWebExtensionWindow openNewTabAtIndex:]): Use class type.
(-[TestWebExtensionWindow closeTab:]): Call closeTab:windowIsClosing:.
(-[TestWebExtensionWindow closeTab:windowIsClosing:]): Added. Close the web view. Update activeTab if needed.
(-[TestWebExtensionWindow replaceTab:withTab:]): Close the web view on the replaced tab.
(-[TestWebExtensionWindow moveTab:toIndex:]): Use class type.

Canonical link: https://commits.webkit.org/269087@main
